cli_tools/google-osconfig-agent/ospatch/patch_run.go

```diff
@@ -277,70 +279,54 @@ func (r *patchRun) reportSucceeded(ctx context.Context) {
  * - An error occurred and we do another attempt starting where we last failed.
  * - The process was unexpectedly restarted and we are continuing from where we left off.
  */
-func (r *patchRun) runPatch(ctx context.Context) {
-	savedPatchState, err := loadState(state)
-	if err != nil {
-		logger.Errorf("loadState error: %v", err)
-		return
+func (r *patchRun) runPatch() {
+	logger.Debugf("Running patch job %s", r.Job.PatchJob)
+	if err := r.createClient(); err != nil {
+		logger.Errorf("Error creating osconfig client: %v", err)
 	}
+	defer r.close()
 
-	if savedPatchState != nil && savedPatchState.Job.PatchJob == r.Job.PatchJob {
-		// continue from previous patch step
-		r.PatchStep = savedPatchState.PatchStep
-	} else {
-		// We have no saved state for this patch.
-		r.PatchStep = started
-	}
-
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[started] {
-		logger.Debugf("Starting patchJob %s", r.Job)
-		r.StartedAt = time.Now()
-
-		if r.reportState(ctx, osconfigpb.Instance_STARTED) {
+	for {
+		switch r.PatchStep {
+		default:
+			r.finishAndReportError(fmt.Sprintf("Unknown step: %q", r.PatchStep))
 			return
-		}
-	}
-
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[prePatchReboot] {
-		// check if we need to reboot
-		if r.rebootIfNeeded(ctx, false) {
-			return
-		}
-	}
+		case acked:
+			logger.Debugf("Starting patchJob %s", r.Job)
+			r.StartedAt = time.Now()
+			r.PatchStep = prePatchReboot
 
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[applyPatch] {
-		if r.Job.DryRun {
-			if r.reportState(ctx, osconfigpb.Instance_APPLYING_PATCHES) {
+			if r.reportState(osconfigpb.Instance_STARTED) {
 				return
 			}
-			logger.Infof("Dry run - No updates applied for patch job '%s'", r.Job.PatchJob)
-		} else {
-			err = runUpdates(ctx, r)
-			if err != nil {
-				r.finishAndReportError(ctx, fmt.Sprintf("Failed to apply patches: %v", err))
+		case prePatchReboot:
+			if r.prePatchReboot() {
 				return
 			}
-		}
-	}
-
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[postPatchReboot] {
-		// check if we need to reboot
-		if r.rebootIfNeeded(ctx, true) {
+			r.PatchStep = patching
```

Comment:
Since prePatchReboot() will never return if it's rebooting, don't you need to change the step first?

Reply:
It will loop until a reboot is not needed right now

Reply:
Might be good to add a comment to that effect.